### PR TITLE
Paniukhin HW07

### PR DESCRIPTION
**Pull request by npanuhin**

### CODE REVIEW

**Comment by kajigor** [comment]
2: no solution

File: monads/test/Spec.hs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+evalDirect :: Expr String -> M.Map String Int -> Either (Error String) Int
+evalDirect (Expr.Num n) _ = Right n
+evalDirect (Expr.Var v) env =
+  case M.lookup v env of
+    Just val -> Right val
+    Nothing  -> Left (VarUndefined v)
+evalDirect (Expr.Plus e1 e2) env = do
+  val1 <- evalDirect e1 env
+  val2 <- evalDirect e2 env
+  return (val1 + val2)
```
Comment (kajigor) [inline monads/test/Spec.hs]: What about `Let`?

**Review by kajigor — CHANGES_REQUESTED** [review]

**Comment by kajigor** [comment]
You have a bunch of really good properties, but could you add let-expressions to the generator and the interpreter?

File: monads/test/Spec.hs (hunk 1 above)
Comment (npanuhin) [reply monads/test/Spec.hs]: [Added](https://github.com/kajigor/fp-2024-cub-fall/pull/149/commits/049306b7efbbd2309d30a76fcdee133cbf2f2a45#diff-0e851a1a04a54af327b8044f27cdd1c408a2fee485bb48c4638e5656bdb5e8a3R109-R111)

**Comment by npanuhin** [comment]
Added Let-expressions to `evalDirect` and `prop_directEvaluationConsistency` test

**Comment by kajigor** [comment]
Ok.

**Pull request closed without merging**